mars/dataframe/utils.py

```diff
@@ -1456,23 +1455,91 @@ def clean_up_func(op):
             )
             return
 
-    func = op.func
-    if hasattr(func, "__closure__") and func.__closure__ is not None:
-        counted_bytes = 0
-        for cell in func.__closure__:
-            counted_bytes += sys.getsizeof(cell.cell_contents)
-            if counted_bytes >= closure_clean_up_bytes_threshold:
-                op.need_clean_up_func = True
-                break
     # Note: op.func_key is set only when op.need_clean_up_func is True.
-    if op.need_clean_up_func:
+    if whether_to_clean_up(op, threshold) is True:
         assert (
             op.logic_key is not None
         ), "Logic key wasn't calculated before cleaning up func."
         op.func_key = ctx.storage_put(op.func)
         op.func = None
 
 
+def whether_to_clean_up(op, threshold):
+    func = op.func
+    counted_bytes = 0
+    max_recursion_depth = 2
+
+    from numbers import Number
+    from collections import deque
+
+    BYPASS_CLASSES = (str, bytes, Number, range, bytearray, pd.DataFrame, pd.Series)
+
+    def check_exceed_threshold():
+        nonlocal threshold, counted_bytes
+        if counted_bytes >= threshold:
+            raise StopIteration
```

Comment:
The `StopIteration` has special meanings. If it is raised inside an iteration loop, then the loop will be stopped. Use a `RuntimeError` or just the `Exception`.